File: examples/framework/python/mpc_examples_framework/waypoints_simulator.py
#!/usr/bin/env python3

# Copyright 2025 Universidad Politécnica de Madrid
# SPDX-License-Identifier: BSD-3-Clause

"""Triple-loop waypoint simulator driving an :class:`IController` with an
:class:`ITrajectoryGenerator`.

Python mirror of ``examples/framework/src/waypoints_simulator.cpp``. Delegates
physics, INDI and IMU updates to :class:`mavpy.simulator.Simulator` in
``RATES`` control mode and logs data via ``mpc_acados_core.logging.CsvLogger``.
"""

__authors__ = 'Rafael Perez-Segui'
__license__ = 'BSD-3-Clause'

import sys
import time
from dataclasses import dataclass
from typing import List, Optional

import numpy as np
from mavpy.controllers import ControlMode
from mavpy.simulator import Simulator, SimulatorParameters
from mpc_acados_core.logging.csv_logger import CsvLogger, euler_to_quaternion

from .controller_base import IController
from .example_config import ExampleConfig
from .trajectory_generator_base import ITrajectoryGenerator
from .types import ReferenceField, ReferenceSample, has_field


@dataclass
class BenchmarkStats:
    """Aggregate timing statistics produced by one :meth:`WaypointsSimulator.run`."""

    simulated_time_s: float = 0.0
    real_time_s: float = 0.0
    sim_speedup: float = 0.0
    controller_mean_us: float = 0.0
    indi_mean_us: float = 0.0
    imu_mean_us: float = 0.0
    model_mean_us: float = 0.0
    controller_steps: int = 0
    indi_steps: int = 0


def _describe_fields_mask(mask: ReferenceField) -> str:
    parts = []
    if has_field(mask, ReferenceField.POSITION):
        parts.append('position')
    if has_field(mask, ReferenceField.VELOCITY):
        parts.append('velocity')
    if has_field(mask, ReferenceField.ACCELERATION):
        parts.append('acceleration')
    return ', '.join(parts) if parts else '<none>'


def _is_finite_vec3(v: np.ndarray) -> bool:
    return bool(np.all(np.isfinite(v)))


def _mean_of_seconds_in_us(seconds: List[float]) -> float:
    if not seconds:
        return 0.0
    return (sum(seconds) / len(seconds)) * 1e6


def _print_progress(fraction: float) -> None:
    fraction = max(0.0, min(1.0, fraction))
    bar_width = 40
    filled = int(bar_width * fraction)
    bar = '#' * filled + '-' * (bar_width - filled)
    sys.stdout.write(f'\r[{bar}] {fraction * 100.0:6.2f}%')
    sys.stdout.flush()


class WaypointsSimulator:
    """Orchestrates the triple-loop waypoint simulation.

    Construction takes ownership of the controller and the generator and pairs
    them with an ``mavpy.Simulator``. A compatibility warning is emitted once
    in the constructor if the generator does not cover every
    :class:`ReferenceField` required by the controller; the run proceeds
    anyway with missing fields held at zero (useful to study controller
    degradation under incomplete references).
    """

    def __init__(
        self,
        controller: IController,
        traj_gen: ITrajectoryGenerator,
        example_cfg: ExampleConfig,
        simulator_params: SimulatorParameters,
        output_csv: str,
    ) -> None:
        if controller is None:
            raise ValueError('WaypointsSimulator: controller must not be None.')
        if traj_gen is None:
            raise ValueError('WaypointsSimulator: trajectory generator must not be None.')
        if not example_cfg.waypoints:
            raise ValueError('WaypointsSimulator: example_cfg.waypoints is empty.')

        self._controller = controller
        self._traj_gen = traj_gen
        self._example_cfg = example_cfg
        self._output_csv = output_csv
        self._sim = Simulator(simulator_params)
        self._stats = BenchmarkStats()

        self._check_compatibility()

    def _check_compatibility(self) -> None:
        required = self._controller.required_reference_fields()
        provided = self._traj_gen.provided_reference_fields()
        missing = ReferenceField(required & ~provided)
        if missing != ReferenceField.NONE:
            sys.stderr.write(
                f"[WaypointsSimulator] Warning: generator '{self._traj_gen.name()}' "
                f"does not produce field(s) required by controller "
                f"'{self._controller.name()}': {_describe_fields_mask(missing)}. "
                'Missing fields will be held at zero; controller may not '
                'converge as expected.\n'
            )

    @property
    def stats(self) -> BenchmarkStats:
        """Return the benchmark statistics collected by the last :meth:`run`."""
        return self._stats

    def run(self) -> None:
        """Execute the mission until ``sim_time + hover_time`` elapses."""
        sim = self._sim
        sim.arm()
        sim.set_control_mode(ControlMode.RATES)

        initial_state = sim.state
        self._controller.initialize(initial_state, self._example_cfg)
        self._traj_gen.initialize(
            self._example_cfg.waypoints, initial_state, self._example_cfg)

        model_dt = self._example_cfg.model_dt
        controller_dt = self._example_cfg.controller_dt
        outer_dt = self._controller.control_period()
        hover_time = self._example_cfg.hover_time
        max_sim_time = self._example_cfg.sim_time + hover_time
        benchmark = self._example_cfg.benchmark
        silent = self._example_cfg.silent
        max_speed = self._example_cfg.max_speed

        n_samples = self._controller.reference_horizon_size()
        dt_h = self._controller.reference_horizon_dt()
        refs: List[ReferenceSample] = [ReferenceSample() for _ in range(n_samples)]

        logger: Optional[CsvLogger] = None
        if not benchmark:
            logger = CsvLogger(self._output_csv)

        controller_times: List[float] = []
        indi_times: List[float] = []
        imu_times: List[float] = []
        model_times: List[float] = []

        prefix = '[BENCHMARK] ' if benchmark else ''
        print(
            f'{prefix}Controller   : {self._controller.name()}\n'
            f'Generator    : {self._traj_gen.name()}\n'
            f'Output file  : {self._output_csv}\n'
            f'Waypoints    : {len(self._example_cfg.waypoints)}\n'
            f'Horizon      : N={n_samples}, dt_h={dt_h} s\n'
            f'Outer period : {outer_dt} s\n'
            f'Max sim time : {max_sim_time} s\n'
        )
        print(
            f'Waypoint 1/{len(self._example_cfg.waypoints)}: '
            f'{self._example_cfg.waypoints[0]}'
        )

        zero3 = np.zeros(3)
        identity_quat = np.array([1.0, 0.0, 0.0, 0.0])
        zero4 = np.zeros(4)

        if logger is not None:
            logger.save(
                0.0, zero3, identity_quat, zero3, zero3, zero3, identity_quat,
                0.0, zero3, zero4, 0.0, 0, False, max_speed,
            )

        ref0_snapshot = ReferenceSample()
        hover_active = False
        hover_end_time = max_sim_time
        last_wp_index = self._traj_gen.current_waypoint_index()

        wall_start = time.perf_counter()
        t = 0.0

        while t < max_sim_time + 1e-9:
            if hover_active and t >= hover_end_time - 1e-9:
                break

            state = sim.state
            position = np.asarray(state.position, dtype=float)
            orientation = np.asarray(state.orientation, dtype=float)
            velocity = np.asarray(state.linear_velocity, dtype=float)

            if (not _is_finite_vec3(position)
                    or not bool(np.all(np.isfinite(orientation)))
                    or not _is_finite_vec3(velocity)):
                sys.stderr.write(f'Non-finite state at t={t} s\n')
                break

            self._traj_gen.update(t, state)
            for k in range(n_samples):
                refs[k] = self._traj_gen.evaluate(t + k * dt_h)
            ref0_snapshot = refs[0]

            cmd = self._controller.compute_command(state, refs)
            solve_us = self._controller.last_solve_time_micros()
            controller_times.append(solve_us * 1e-6)

            current_wp = self._traj_gen.current_waypoint_index()
            if current_wp != last_wp_index:
                if not silent:
                    print(
                        f'\nWaypoint {current_wp + 1}/{len(self._example_cfg.waypoints)}: '
                        f'{self._example_cfg.waypoints[current_wp]}'
                    )
                last_wp_index = current_wp

            if not hover_active and self._traj_gen.is_finished(t):
                hover_active = True
                hover_end_time = t + hover_time
                if not silent:
                    print(f'\nMission finished. Hovering for {hover_time} s')

            ref_orientation = euler_to_quaternion(0.0, 0.0, ref0_snapshot.yaw)
            sim.set_reference_rates(float(cmd.thrust_n),
                                    np.asarray(cmd.angular_rate, dtype=float))

            t_ctrl = t
            while t_ctrl < t + outer_dt - 1e-9:
                indi_t0 = time.perf_counter()
                sim.update_controller(controller_dt)
                indi_t1 = time.perf_counter()

                sim.update_imu(controller_dt)
                indi_t2 = time.perf_counter()

                t_model = t_ctrl
                while t_model < t_ctrl + controller_dt - 1e-9:
                    sim.update_model(model_dt)
                    t_model += model_dt
                indi_t3 = time.perf_counter()

                indi_times.append(indi_t1 - indi_t0)
                imu_times.append(indi_t2 - indi_t1)
                model_times.append(indi_t3 - indi_t2)

                t_ctrl += controller_dt

                if logger is not None:
                    s = sim.state
                    logger.save(
                        t_ctrl,
                        np.asarray(s.position, dtype=float),
                        np.asarray(s.orientation, dtype=float),
                        np.asarray(s.linear_velocity, dtype=float),
                        np.asarray(s.angular_velocity, dtype=float),
                        np.asarray(ref0_snapshot.position, dtype=float),
                        np.asarray(ref_orientation, dtype=float),
                        float(cmd.thrust_n),
                        np.asarray(cmd.angular_rate, dtype=float),
                        np.asarray(s.motor_angular_velocity, dtype=float),
                        float(solve_us),
                        int(self._traj_gen.current_waypoint_index()),
                        bool(hover_active),
                        float(max_speed),
                    )

            t += outer_dt
            if not silent:
                _print_progress(t / max_sim_time)

        wall_end = time.perf_counter()
        real_time_s = wall_end - wall_start

        self._stats = BenchmarkStats(
            simulated_time_s=t,
            real_time_s=real_time_s,
            sim_speedup=(t / real_time_s) if real_time_s > 0.0 else 0.0,
            controller_mean_us=_mean_of_seconds_in_us(controller_times),
            indi_mean_us=_mean_of_seconds_in_us(indi_times),
            imu_mean_us=_mean_of_seconds_in_us(imu_times),
            model_mean_us=_mean_of_seconds_in_us(model_times),
            controller_steps=len(controller_times),
            indi_steps=len(indi_times),
        )

        if logger is not None:
            logger.close()

    def print_benchmark(self) -> None:
        """Pretty-print the :class:`BenchmarkStats` from the last run."""
        s = self._stats
        print(
            f'\n\nSimulated time      : {s.simulated_time_s} s\n'
            f'Real time           : {s.real_time_s} s\n'
            f'Sim speedup         : {s.sim_speedup}x\n'
            f'Controller compute  : {s.controller_mean_us} us (avg, '
            f'{s.controller_steps} steps)\n'
            f'INDI controller     : {s.indi_mean_us} us (avg)\n'
            f'IMU update          : {s.imu_mean_us} us (avg)\n'
            f'Physics model       : {s.model_mean_us} us (avg, per INDI step)'
        )

File: examples/framework/src/waypoints_simulator.cpp
// Copyright 2025 Universidad Politécnica de Madrid
//
// Redistribution and use in source and binary forms, with or without
// modification, are permitted provided that the following conditions are met:
//
//    * Redistributions of source code must retain the above copyright
//      notice, this list of conditions and the following disclaimer.
//
//    * Redistributions in binary form must reproduce the above copyright
//      notice, this list of conditions and the following disclaimer in the
//      documentation and/or other materials provided with the distribution.
//
//    * Neither the name of the Universidad Politécnica de Madrid nor the
//      names of its contributors may be used to endorse or promote products
//      derived from this software without specific prior written permission.
//
// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
// POSSIBILITY OF SUCH DAMAGE.

/**
 * @file waypoints_simulator.cpp
 * @author Rafael Perez-Segui <r.psegui@upm.es>
 */

#include "framework/waypoints_simulator.hpp"

#include <Eigen/Dense>

#include <chrono>
#include <cmath>
#include <cstddef>
#include <iostream>
#include <memory>
#include <stdexcept>
#include <string>
#include <vector>

#include "framework/types.hpp"
#include "utils/utils.hpp"

namespace mpc_examples::framework {

namespace {

std::string describeFieldsMask(ReferenceFieldMask mask) {
  std::string out;
  auto append = [&](const char* name, ReferenceField f) {
    if (hasField(mask, f)) {
      if (!out.empty()) out += ", ";
      out += name;
    }
  };
  append("position", ReferenceField::kPosition);
  append("velocity", ReferenceField::kVelocity);
  append("acceleration", ReferenceField::kAcceleration);
  return out.empty() ? "<none>" : out;
}

bool isFiniteVector3(const Eigen::Vector3d& v) {
  return std::isfinite(v.x()) && std::isfinite(v.y()) && std::isfinite(v.z());
}

bool isFiniteQuaternion(const Eigen::Quaterniond& q) {
  return std::isfinite(q.w()) && std::isfinite(q.x()) && std::isfinite(q.y()) &&
         std::isfinite(q.z());
}

double meanOfTimesUs(const std::vector<double>& seconds) {
  if (seconds.empty()) {
    return 0.0;
  }
  double sum = 0.0;
  for (double x : seconds) {
    sum += x;
  }
  return sum / static_cast<double>(seconds.size()) * 1e6;
}

}  // namespace

WaypointsSimulator::WaypointsSimulator(
    std::unique_ptr<IController> controller,
    std::unique_ptr<ITrajectoryGenerator> traj_gen,
    const ExampleConfig& example_cfg,
    const mav_simulator::SimulatorParameters& simulator_params,
    const std::string& output_csv)
    : controller_(std::move(controller)),
      traj_gen_(std::move(traj_gen)),
      example_cfg_(example_cfg),
      output_csv_(output_csv),
      sim_(simulator_params) {
  if (!controller_) {
    throw std::invalid_argument("WaypointsSimulator: controller must not be null.");
  }
  if (!traj_gen_) {
    throw std::invalid_argument("WaypointsSimulator: trajectory generator must not be null.");
  }
  if (example_cfg_.waypoints.empty()) {
    throw std::invalid_argument("WaypointsSimulator: example_cfg.waypoints is empty.");
  }

  checkCompatibility_();
}

void WaypointsSimulator::checkCompatibility_() const {
  const ReferenceFieldMask required = controller_->requiredReferenceFields();
  const ReferenceFieldMask provided = traj_gen_->providedReferenceFields();
  const ReferenceFieldMask missing  = required & ~provided;
  if (missing != 0u) {
    std::cerr << "[WaypointsSimulator] Warning: generator '" << traj_gen_->name()
              << "' does not produce field(s) required by controller '" << controller_->name()
              << "': " << describeFieldsMask(missing)
              << ". Missing fields will be held at zero; controller may not converge as expected."
              << std::endl;
  }
}

void WaypointsSimulator::run() {
  // --- Simulator setup ------------------------------------------------------
  sim_.arm();
  sim_.setControlMode(mav_simulator::ControlMode::RATES);

  const mav_model::State& initial_state = sim_.getState();
  controller_->initialize(initial_state, example_cfg_);
  traj_gen_->initialize(example_cfg_.waypoints, initial_state, example_cfg_);

  // --- Timing parameters ----------------------------------------------------
  const double model_dt      = example_cfg_.model_dt;
  const double controller_dt = example_cfg_.controller_dt;
  const double outer_dt      = controller_->controlPeriod();
  const double hover_time    = example_cfg_.hover_time;
  const double max_sim_time  = example_cfg_.sim_time + hover_time;
  const bool benchmark       = example_cfg_.benchmark;
  const bool silent          = example_cfg_.silent;
  const double max_speed     = example_cfg_.max_speed;

  const int N_samples   = controller_->referenceHorizonSize();
  const double dt_h     = controller_->referenceHorizonDt();
  std::vector<ReferenceSample> refs(static_cast<std::size_t>(N_samples));

  // --- Logging --------------------------------------------------------------
  std::unique_ptr<CsvLogger> logger;
  if (!benchmark) {
    logger = std::make_unique<CsvLogger>(output_csv_);
  }

  // --- Benchmark buffers ----------------------------------------------------
  const auto est_outer_steps = static_cast<std::size_t>(max_sim_time / outer_dt) + 1U;
  const auto est_indi_steps  = static_cast<std::size_t>(max_sim_time / controller_dt) + 1U;
  std::vector<double> controller_times;
  std::vector<double> indi_times;
  std::vector<double> imu_times;
  std::vector<double> model_times;
  controller_times.reserve(est_outer_steps);
  indi_times.reserve(est_indi_steps);
  imu_times.reserve(est_indi_steps);
  model_times.reserve(est_indi_steps);

  std::cout << (benchmark ? "[BENCHMARK] " : "") << "Controller   : " << controller_->name()
            << "\n"
            << "Generator    : " << traj_gen_->name() << "\n"
            << "Output file  : " << output_csv_ << "\n"
            << "Waypoints    : " << example_cfg_.waypoints.size() << "\n"
            << "Horizon      : N=" << N_samples << ", dt_h=" << dt_h << " s\n"
            << "Outer period : " << outer_dt << " s\n"
            << "Max sim time : " << max_sim_time << " s\n\n";
  std::cout << "Waypoint 1/" << example_cfg_.waypoints.size() << ": "
            << example_cfg_.waypoints.front().transpose() << "\n";

  // Initial log entry at t=0 with zero references (pre-arm snapshot).
  if (logger) {
    logger->save(0.0, Eigen::Vector3d::Zero(), Eigen::Quaterniond::Identity(),
                 Eigen::Vector3d::Zero(), Eigen::Vector3d::Zero(), Eigen::Vector3d::Zero(),
                 Eigen::Quaterniond::Identity(), 0.0, Eigen::Vector3d::Zero(),
                 Eigen::Matrix<double, 4, 1>::Zero(), 0.0, 0, false, max_speed);
  }

  // --- Main loop ------------------------------------------------------------
  ControlCommand cmd;
  ReferenceSample ref0_snapshot;
  bool hover_active     = false;
  double hover_end_time = max_sim_time;
  int last_wp_index     = traj_gen_->currentWaypointIndex();

  const auto wall_start = std::chrono::high_resolution_clock::now();
  double t              = 0.0;

  while (t < max_sim_time + 1e-9) {
    if (hover_active && t >= hover_end_time - 1e-9) {
      break;
    }

    // Read current ground-truth state
    const mav_model::State& state        = sim_.getState();
    const Eigen::Vector3d position       = state.getPositionVector();
    const Eigen::Quaterniond orientation = state.getOrientationVector();
    const Eigen::Vector3d velocity       = state.getLinearVelocityVector();

    if (!isFiniteVector3(position) || !isFiniteQuaternion(orientation) ||
        !isFiniteVector3(velocity)) {
      std::cerr << "Non-finite state at t=" << t << " s\n";
      break;
    }

    // Refresh generator and sample the controller horizon
    traj_gen_->update(t, state);
    for (int k = 0; k < N_samples; ++k) {
      refs[static_cast<std::size_t>(k)] = traj_gen_->evaluate(t + k * dt_h);
    }
    ref0_snapshot = refs.front();

    // Compute control command
    cmd = controller_->computeCommand(state, refs);
    const double solve_us = controller_->lastSolveTimeMicros();
    controller_times.push_back(solve_us * 1e-6);

    // Notify waypoint advances (generator-driven)
    const int current_wp = traj_gen_->currentWaypointIndex();
    if (current_wp != last_wp_index) {
      if (!silent) {
        std::cout << "\nWaypoint " << current_wp + 1 << "/" << example_cfg_.waypoints.size()
                  << ": " << example_cfg_.waypoints[static_cast<std::size_t>(current_wp)].transpose();
      }
      last_wp_index = current_wp;
    }

    // Transition to hover once the generator reports the mission finished
    if (!hover_active && traj_gen_->isFinished(t)) {
      hover_active   = true;
      hover_end_time = t + hover_time;
      if (!silent) {
        std::cout << "\nMission finished. Hovering for " << hover_time << " s";
      }
    }

    const Eigen::Quaterniond ref_orientation = eulerToQuaternion(0.0, 0.0, ref0_snapshot.yaw);

    // --- INDI + physics inner loop (zero-order hold on outer command) -----
    sim_.setReferenceRates(cmd.thrust_n, cmd.angular_rate);

    double t_ctrl = t;
    while (t_ctrl < t + outer_dt - 1e-9) {
      const auto indi_t0 = std::chrono::high_resolution_clock::now();
      sim_.updateController(controller_dt);
      const auto indi_t1 = std::chrono::high_resolution_clock::now();

      sim_.updateImu(controller_dt);
      const auto indi_t2 = std::chrono::high_resolution_clock::now();

      double t_model = t_ctrl;
      while (t_model < t_ctrl + controller_dt - 1e-9) {
        sim_.updateModel(model_dt);
        t_model += model_dt;
      }
      const auto indi_t3 = std::chrono::high_resolution_clock::now();

      indi_times.push_back(std::chrono::duration<double>(indi_t1 - indi_t0).count());
      imu_times.push_back(std::chrono::duration<double>(indi_t2 - indi_t1).count());
      model_times.push_back(std::chrono::duration<double>(indi_t3 - indi_t2).count());

      t_ctrl += controller_dt;

      if (logger) {
        const mav_model::State& s = sim_.getState();
        logger->save(t_ctrl, s.getPositionVector(), s.getOrientationVector(),
                     s.getLinearVelocityVector(), s.getAngularVelocityVector(),
                     ref0_snapshot.position, ref_orientation, cmd.thrust_n, cmd.angular_rate,
                     s.getMotorAngularVelocityVector(), solve_us,
                     traj_gen_->currentWaypointIndex(), hover_active, max_speed);
      }
    }

    t += outer_dt;
    if (!silent) {
      printProgress(t / max_sim_time);
    }
  }

  const auto wall_end = std::chrono::high_resolution_clock::now();
  const double real_time_s = std::chrono::duration<double>(wall_end - wall_start).count();

  stats_ = BenchmarkStats{
      .simulated_time_s   = t,
      .real_time_s        = real_time_s,
      .sim_speedup        = (real_time_s > 0.0) ? (t / real_time_s) : 0.0,
      .controller_mean_us = meanOfTimesUs(controller_times),
      .indi_mean_us       = meanOfTimesUs(indi_times),
      .imu_mean_us        = meanOfTimesUs(imu_times),
      .model_mean_us      = meanOfTimesUs(model_times),
      .controller_steps   = controller_times.size(),
      .indi_steps         = indi_times.size(),
  };
}

void WaypointsSimulator::printBenchmark() const {
  std::cout << "\n\nSimulated time      : " << stats_.simulated_time_s << " s\n"
            << "Real time           : " << stats_.real_time_s << " s\n"
            << "Sim speedup         : " << stats_.sim_speedup << "x\n"
            << "Controller compute  : " << stats_.controller_mean_us << " µs (avg, "
            << stats_.controller_steps << " steps)\n"
            << "INDI controller     : " << stats_.indi_mean_us << " µs (avg)\n"
            << "IMU update          : " << stats_.imu_mean_us << " µs (avg)\n"
            << "Physics model       : " << stats_.model_mean_us << " µs (avg, per INDI step)\n";
}

}  // namespace mpc_examples::framework
